internal/rule/squashjoins.go

```diff
@@ -0,0 +1,948 @@
+	n, err := n.TransformUp(func(n sql.Node) (sql.Node, error) {
+		join, ok := n.(*plan.InnerJoin)
+		if !ok {
+			return n, nil
+		}
+
+		return squashJoin(join)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return n.TransformUp(func(n sql.Node) (sql.Node, error) {
+		t, ok := n.(*joinedTables)
+		if !ok {
+			return n, nil
+		}
+
+		return buildSquashedTable(t.tables, t.filters, t.columns)
+	})
+}
+
+func squashJoin(join *plan.InnerJoin) (sql.Node, error) {
+	if !isJoinSquashable(join) {
+		return join, nil
+	}
+
+	table, err := joinTables(join)
+	if err != nil {
+		return nil, err
+	}
+
+	return rearrange(join, table), nil
+}
+
+func joinTables(join *plan.InnerJoin) (sql.Table, error) {
+	var tables []sql.Table
+	var filters []sql.Expression
+	var columns []sql.Expression
+	plan.Inspect(join, func(node sql.Node) bool {
+		switch node := node.(type) {
+		case *joinedTables:
+			tables = append(tables, node.tables...)
+			columns = append(columns, node.columns...)
+			filters = append(filters, node.filters...)
+		case *plan.PushdownProjectionAndFiltersTable:
+			table, ok := node.PushdownProjectionAndFiltersTable.(gitquery.Table)
+			if ok {
+				filters = append(filters, node.Filters...)
+				columns = append(columns, node.Columns...)
+				tables = append(tables, table)
+			}
+		case *plan.InnerJoin:
+			filters = append(filters, exprToFilters(node.Cond)...)
+		}
+		return true
+	})
+
+	return &joinedTables{
+		tables:  tables,
+		filters: filters,
+		columns: columns,
+	}, nil
+}
+
+func rearrange(join *plan.InnerJoin, squashedTable sql.Table) sql.Node {
+	var projections []sql.Expression
+	var filters []sql.Expression
+	plan.Inspect(join, func(node sql.Node) bool {
+		switch node := node.(type) {
+		case *plan.Project:
+			projections = append(projections, node.Projections...)
+		case *plan.Filter:
+			filters = append(filters, node.Expression)
+		}
+		return true
+	})
+
+	var node sql.Node = squashedTable
+	if len(filters) > 0 {
+		node = plan.NewFilter(expression.JoinAnd(filters...), node)
+	}
+
+	if len(projections) > 0 {
+		node = plan.NewProject(projections, node)
+	}
+
+	return node
+}
+
+var errInvalidIteratorChain = errors.NewKind("invalid iterator to chain with %s: %T")
+
+func buildSquashedTable(
+	tables []sql.Table,
+	filters, columns []sql.Expression,
+) (sql.Node, error) {
+	tableNames := orderedTableNames(tables)
+
+	var iter gitquery.ChainableIter
+	var err error
+	for _, t := range tableNames {
+		switch t {
+		case "repositories":
+			switch iter.(type) {
+			case nil:
+				var f sql.Expression
+				f, filters, err = filtersForTable(
+					gitquery.RepositoriesTableName,
+					filters,
+					gitquery.RepositoriesSchema,
+				)
+				if err != nil {
+					return nil, err
+				}
+				iter = gitquery.NewAllReposIter(f)
+			default:
+				return nil, errInvalidIteratorChain.New("repositories", iter)
+			}
+		case "remotes":
+			switch it := iter.(type) {
+			case gitquery.ReposIter:
+				var f sql.Expression
```

Comment:
I don't know if it's worth it or just for code clarity, but this logic is almost duplicated on `case "refs"`

Reply:
Just the same comment for all the duplicated cases.

Reply:
Yep, but there's no way to do that generically without repeating some code because the types and the params of everything inside those cases are not the same.